In [1]:
# Imports and pip installations (if needed)
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder, StandardScaler

# Part 1: Load the dataset

In [3]:
# Load the given datasets

df1 = pd.read_csv('/workspaces/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_categorical.csv')
df2 = pd.read_csv('/workspaces/assignment-3-data-cleaning-and-preprocessing-<redacted>/chronic_kidney_disease_numerical.csv')



# Print the data
print(df1.head())
print(df2.head())


   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     224481  3.0  NaN       NaN       NaN  notpresent  notpresent  yes  yes   
1     992643  0.0  0.0       NaN       NaN  notpresent  notpresent   no   no   
2     308740  1.0  0.0  abnormal  abnormal  notpresent  notpresent  yes   no   
3     450314  1.0  0.0       NaN    normal  notpresent  notpresent   no   no   
4     881763  0.0  0.0    normal    normal  notpresent  notpresent   no   no   

  cad appet   pe  ane  Target  
0  no  good  yes   no     ckd  
1  no  good   no   no  notckd  
2  no  good  yes   no     ckd  
3  no  good   no  yes     ckd  
4  no  good   no   no  notckd  
   unique_id   age    bp    bgr     bu   sc    sod  pot  hemo   pcv    wbcc  \
0     203694  38.0  80.0   99.0   19.0  0.5  147.0  3.5  13.6  44.0  7300.0   
1     938027  43.0  60.0  108.0   25.0  1.0  144.0  5.0  17.8  43.0  7200.0   
2     421471  37.0  60.0  111.0   35.0  0.8  135.0  4.1  16.2  50.0  5500.0   
3     76411

# Part 2: Analyze the Dataset

Refer to this: https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease

Explain what the each data is in your own words. What are the features and labels? Are the features in the given datasets : categorical, numerical or both? Give 3 examples of categorical and numerical columns each (if they exist)

Answer: The Chronic Kidney Disease (CKD) dataset from the UCI Machine Learning Repository contains data about patients, which is used to classify whether or not they have chronic kidney disease. Each row in the dataset represents an individual patient, and each column represents a specific feature (attribute) or label (target variable) related to their medical condition or test results.

1) Features: These are the input variables that describe patient characteristics and medical test results. In this dataset, features include patient information such as age, blood pressure, specific blood test results, and other medical indicators.

2) Label (Target): This is the output variable that the dataset is trying to predict, which indicates whether the patient has chronic kidney disease (CKD) or not. The label is typically denoted as "class," and it can take two values: "ckd" (has chronic kidney disease) or "notckd" (does not have chronic kidney disease).

The features in the dataset are a mix of categorical (qualitative) and numerical (quantitative) data.

Examples of Categorical Features:
1) "RBC" (Red Blood Cells): Can take values like "normal" or "abnormal".
2) "Pus_cell" (Pus Cells): Indicates "normal" or "abnormal".
3) "Class": This is the label itself and is a categorical feature with two possible values: "ckd" (chronic kidney disease) or "notckd" (no chronic kidney disease).

Examples of Numerical Features:
1) "Age": Represents the patient's age in years (continuous numerical value).
2) "Blood_Pressure": Measures blood pressure in mmHg (numerical, continuous).
3) "Serum_creatinine": The concentration of creatinine in the blood, in mg/dL (numerical, continuous).

# Part 3: Data Preprocessing

A fundamental skill in Machine Learning is mastering the art of data cleaning and preprocessing. In this assignment, you will learn and apply essential data cleaning techniques to transform a raw dataset into a clean, ready-to-use form which you can use for regression or classification tasks. By the end of this assignment, you'll have a fully clean dataset and a solid foundation in preparing data for various machine learning models.

## Part 3.1 : Drop Duplicate rows

Let's start by checking if the given datasets have any duplicate rows (same Unique Id). Use pandas to identify and remove these duplicate rows from the given dataset

In [4]:
# For the numerical dataset, check if there are duplicate rows in the dataset. If yes, print total number of duplicate rows

numerical_cols1 = df1.select_dtypes(include=['int64', 'float64']).columns
numerical_cols2 = df2.select_dtypes(include=['int64', 'float64']).columns

numerical_duplicates_df1 = df1[numerical_cols1].duplicated()
numerical_duplicates_df2 = df2[numerical_cols2].duplicated()

print(f"Total duplicate rows in numerical dataset 1: {numerical_duplicates_df1.sum()}")
print(f"Total duplicate rows in numerical dataset 2: {numerical_duplicates_df2.sum()}")


# Drop these duplicate rows

df1_numerical_cleaned = df1[~numerical_duplicates_df1]
df2_numerical_cleaned = df2[~numerical_duplicates_df2]

# Confirm that duplicates are removed
print(f"Dataset 1 shape after removing duplicates: {df1_numerical_cleaned.shape}")
print(f"Dataset 2 shape after removing duplicates: {df2_numerical_cleaned.shape}")



# Repeat the same for categorical dataset. Print the duplicate rows and drop them

# For the categorical dataset, check if there are duplicate rows
# First, identify the categorical columns
categorical_cols1 = df1.select_dtypes(include=['object']).columns
categorical_cols2 = df2.select_dtypes(include=['object']).columns

# Check for duplicate rows in the categorical dataset
categorical_duplicates_df1 = df1[categorical_cols1].duplicated(keep='first')
categorical_duplicates_df2 = df2[categorical_cols2].duplicated(keep='first')

# Print the total number of duplicate rows
print(f"Total duplicate rows in categorical dataset 1: {categorical_duplicates_df1.sum()}")
print(f"Total duplicate rows in categorical dataset 2: {categorical_duplicates_df2.sum()}")

# Drop these duplicate rows for df1
df1_categorical_cleaned = df1[~categorical_duplicates_df1].reset_index(drop=True)

# For df2, since there are no duplicates, we don't need to clean it
df2_categorical_cleaned = df2.copy()

# Confirm that duplicates are removed (df1 only)
print(f"Dataset 1 shape after removing categorical duplicates: {df1_categorical_cleaned.shape}")
print(f"Dataset 2 shape remains unchanged (no categorical duplicates found): {df2_categorical_cleaned.shape}")




Total duplicate rows in numerical dataset 1: 30
Total duplicate rows in numerical dataset 2: 30
Dataset 1 shape after removing duplicates: (400, 14)
Dataset 2 shape after removing duplicates: (390, 12)
Total duplicate rows in categorical dataset 1: 270
Total duplicate rows in categorical dataset 2: 0
Dataset 1 shape after removing categorical duplicates: (160, 14)
Dataset 2 shape remains unchanged (no categorical duplicates found): (420, 12)


## Part 3.2: Combine two differents datasets

A good skill to have is to know how to combine 2 different datasets.

Are all the unique ids are present in both datasets? Why do you think so? If not, what do the rows that are missing from one of the datasets look like in the combined table?

Answer: No, not all unique IDs are present in both datasets. This is likely because the two datasets represent different types of data, collected from different tests or assessments performed on the same or different subsets of patients

In [5]:
# Merge the two given numerical and categorical datasets based on their unique_ID.

combined_df = pd.merge(df1, df2, on='unique_id', how='outer')


#Print the combined dataset

print("Combined Dataset:")
print(combined_df.head())


Combined Dataset:
   unique_id   al   su     rbc        pc         pcc          ba  htn   dm  \
0     102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes   no   
1     103570  0.0  0.0  normal    normal  notpresent  notpresent   no   no   
2     104675  2.0  1.0  normal    normal     present  notpresent   no  yes   
3     105923  1.0  0.0     NaN    normal  notpresent  notpresent  yes  yes   
4     107297  0.0  0.0  normal    normal  notpresent  notpresent   no   no   

  cad  ...    bp    bgr     bu    sc    sod  pot  hemo   pcv     wbcc  rbcc  
0  no  ...  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0   3800.0   2.1  
1  no  ...  80.0  125.0    NaN   NaN  136.0  3.5  15.4  43.0   5600.0   4.5  
2  no  ...  80.0  215.0  133.0   2.5    NaN  NaN  13.2  41.0      NaN   NaN  
3  no  ...  90.0  139.0   89.0   3.0  140.0  4.1  12.0  37.0   7900.0   3.9  
4  no  ...  70.0    NaN    NaN   NaN    NaN  NaN  16.4  43.0  10800.0   5.7  

[5 rows x 25 columns]


## Part 3.3: Rows with Missing values

Removing missing values from a dataset is important for classification because it ensures the model is trained on complete and accurate data, leading to better performance and reliable predictions. Incomplete data can introduce bias and errors, negatively impacting the model's effectiveness.

In [6]:
# Calculate the percentage of rows that contain atleast one missing value

total_rows = combined_df.shape[0]
rows_with_missing_values = combined_df.isnull().any(axis=1).sum()
percentage_missing = (rows_with_missing_values / total_rows) * 100

# Print %
print(f"Percentage of rows with at least one missing value: {percentage_missing:.2f}%")

# Drop these rows from the dataset

cleaned_df = combined_df.dropna()

# Print the Dataset

print("Dataset after dropping rows with missing values:")
print(cleaned_df.head())

Percentage of rows with at least one missing value: 59.76%
Dataset after dropping rows with missing values:
   unique_id   al   su     rbc        pc         pcc          ba  htn  dm cad  \
0     102677  3.0  0.0  normal  abnormal  notpresent  notpresent  yes  no  no   
6     109053  0.0  0.0  normal    normal  notpresent  notpresent   no  no  no   
7     114717  0.0  0.0  normal    normal  notpresent  notpresent   no  no  no   
8     118191  0.0  0.0  normal    normal  notpresent  notpresent   no  no  no   
9     118191  0.0  0.0  normal    normal  notpresent  notpresent   no  no  no   

   ...    bp    bgr     bu    sc    sod  pot  hemo   pcv    wbcc  rbcc  
0  ...  70.0   76.0  186.0  15.0  135.0  7.6   7.1  22.0  3800.0   2.1  
6  ...  60.0  106.0   27.0   0.7  150.0  3.3  14.4  42.0  8100.0   4.7  
7  ...  80.0  122.0   25.0   0.8  138.0  5.0  17.1  41.0  9100.0   5.2  
8  ...  70.0   93.0   32.0   0.9  143.0  4.7  16.6  43.0  7100.0   5.3  
9  ...  70.0   93.0   32.0   0.9  143.0 

## Part 3.4: Sort the dataset according to the Labels

In [7]:
# Sort the dataset according to the values in 'Target' column. Make sure reset the indices after sorting

sorted_df = combined_df.sort_values(by='Target').reset_index(drop=True)


# Print the dataset

print("Dataset sorted by 'Target' column:")
print(sorted_df.head())

Dataset sorted by 'Target' column:
   unique_id   al   su       rbc        pc         pcc          ba  htn   dm  \
0     102677  3.0  0.0    normal  abnormal  notpresent  notpresent  yes   no   
1     619200  0.0  0.0       NaN  abnormal  notpresent  notpresent   no  yes   
2     617089  1.0  3.0    normal  abnormal  notpresent  notpresent   no   no   
3     615697  4.0  0.0  abnormal  abnormal  notpresent     present   no   no   
4     608902  4.0  1.0    normal  abnormal     present  notpresent  yes  yes   

  cad  ...    bp    bgr     bu    sc    sod   pot  hemo   pcv     wbcc  rbcc  
0  no  ...  70.0   76.0  186.0  15.0  135.0   7.6   7.1  22.0   3800.0   2.1  
1  no  ...  80.0  171.0   30.0   1.0    NaN   NaN  13.7  43.0   4900.0   5.2  
2  no  ...  80.0  182.0  391.0  32.0  163.0  39.0   NaN   NaN      NaN   NaN  
3  no  ...  60.0   94.0   67.0   1.0  135.0   4.9   9.9  30.0  16700.0   4.8  
4  no  ...  90.0  176.0  309.0  13.3  124.0   6.5   3.1   9.0   5400.0   2.1  

[5 rows x

## Part 3.5: Encoding Categorical data

In this step, we identify and process the categorical columns in the sorted dataset. We map each unique value in these columns to separate "dummy" columns.

For example, the column 'rbc' will be transformed into two columns 'rbc_normal' and 'rbc_abnormal'. If a row's value in 'rbc' is 'normal', the 'rbc_normal' column will be set to 1 and 'rbc_abnormal' will be set to 0.


**Note: Find a correct pandas function to do this **

In [8]:
# Write code here
categorical_columns = ['rbc', 'pc', 'pcc', 'ba', 'htn', 'dm', 'cad', 'appet', 'pe', 'ane', 'Target']
df_with_dummies = pd.get_dummies(sorted_df, columns=categorical_columns, drop_first=False)

# Print the dataset
print("Dataset with dummy variables for categorical columns:")
print(df_with_dummies.head())

Dataset with dummy variables for categorical columns:
   unique_id   al   su   age    bp    bgr     bu    sc    sod   pot  ...  \
0     102677  3.0  0.0  59.0  70.0   76.0  186.0  15.0  135.0   7.6  ...   
1     619200  0.0  0.0  68.0  80.0  171.0   30.0   1.0    NaN   NaN  ...   
2     617089  1.0  3.0  67.0  80.0  182.0  391.0  32.0  163.0  39.0  ...   
3     615697  4.0  0.0   6.0  60.0   94.0   67.0   1.0  135.0   4.9  ...   
4     608902  4.0  1.0  56.0  90.0  176.0  309.0  13.3  124.0   6.5  ...   

   cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
0    False       False        True  False    True   False     True   
1    False        True       False   True   False    True    False   
2    False        True       False  False    True    True    False   
3    False       False        True   True   False    True    False   
4    False       False        True  False    True   False     True   

   Target_ckd  Target_ckd\t  Target_notckd  
0        True         F

In the example we went through above, another solution is to have a single column for the binary variable. In the downstream modeling would this be equivalent? What effect would this have if the categorical variable could take more than 2 values? For example, let's say we have a categorical feature that is "type of condiment" that can take 5 separate values and we are trying to predict the rating of a particular sandwich.

## Part 3.6 : Remove Outliers from Numerical Columns

Outliers can disproportionately influence the fit of a regression model, potentially leading to a model that does not generalize well therefore it is important that we remove outliers from the numerical columns of the dataset.

For this dataset, we define an outlier to be 3 times the standard deviation from the mean. Drop these outliers from the dataset

In [9]:
# Remove outliers
numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

for col in numerical_columns:
    mean = df_with_dummies[col].mean()
    std = df_with_dummies[col].std()
    
    # Keep rows where the value is within 3 standard deviations from the mean
    df_with_dummies = df_with_dummies[(df_with_dummies[col] >= (mean - 3 * std)) & (df_with_dummies[col] <= (mean + 3 * std))]

# Print the dataset

print("Dataset after removing outliers from numerical columns:")
print(df_with_dummies.head())

Dataset after removing outliers from numerical columns:
    unique_id   al   su   age    bp    bgr     bu    sc    sod  pot  ...  \
16     580048  0.0  4.0  61.0  80.0  360.0   19.0  0.70  137.0  4.4  ...   
20     621332  1.0  0.0  59.0  80.0  303.0   35.0  1.30  122.0  3.5  ...   
27     679864  NaN  NaN  53.0  60.0   91.0  114.0  3.25  142.0  4.3  ...   
38     640519  NaN  NaN  54.0  70.0  111.0  146.0  7.50  141.0  4.7  ...   
40     631648  0.0  0.0  46.0  80.0  160.0   40.0  2.00  140.0  4.1  ...   

    cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
16    False        True       False   True   False    True    False   
20    False       False        True   True   False    True    False   
27    False       False        True  False    True   False     True   
38    False        True       False   True   False    True    False   
40    False       False        True   True   False   False     True   

    Target_ckd  Target_ckd\t  Target_notckd  
16        True

## Part 3.7 : Normalize the Numerical Columns

Normalizing numerical attributes ensures that all features contribute equally to the model by scaling them to a consistent range, which improves model performance and convergence. It prevents features with larger scales from disproportionately influencing the model's learning process.

In [10]:
# Normalize the all Numerical Attributes in the dataset.
from sklearn.preprocessing import MinMaxScaler

numerical_columns = ['age', 'bp', 'bgr', 'bu', 'sc', 'sod', 'pot', 'hemo', 'pcv', 'wbcc', 'rbcc']

# Initialize the MinMaxScaler
scaler = MinMaxScaler()

# Normalize the numerical columns
df_with_dummies[numerical_columns] = scaler.fit_transform(df_with_dummies[numerical_columns])

# Print the dataset

print("Dataset after normalizing numerical columns:")
print(df_with_dummies.head())

Dataset after normalizing numerical columns:
    unique_id   al   su       age    bp       bgr        bu        sc  \
16     580048  0.0  4.0  0.628205  0.50  0.935484  0.058824  0.026087   
20     621332  1.0  0.0  0.602564  0.50  0.751613  0.163399  0.078261   
27     679864  NaN  NaN  0.525641  0.00  0.067742  0.679739  0.247826   
38     640519  NaN  NaN  0.538462  0.25  0.132258  0.888889  0.617391   
40     631648  0.0  0.0  0.435897  0.50  0.290323  0.196078  0.139130   

         sod       pot  ...  cad_yes  appet_good  appet_poor  pe_no  pe_yes  \
16  0.566667  0.500000  ...    False        True       False   True   False   
20  0.066667  0.263158  ...    False       False        True   True   False   
27  0.733333  0.473684  ...    False       False        True  False    True   
38  0.700000  0.578947  ...    False        True       False   True   False   
40  0.666667  0.421053  ...    False       False        True   True   False   

    ane_no  ane_yes  Target_ckd  Target_c

## Part 3.8: Remove Unnecessary columns

Are there any columns in this dataset which are not appropriate for modeling and predictions? Which column(s)? Justify their exclusion and remove them

Answer: Yes, there are certain columns in the dataset that may not be appropriate for modeling and prediction. The main columns that should be considered for exclusion are:

unique_id:
Justification: This is a unique identifier for each patient, and it doesn’t provide any useful information for modeling or predicting Chronic Kidney Disease (CKD). Including this column would only introduce noise, as it has no relationship to the target variable (Target).

In [11]:
#Remove that column
df_cleaned = df_with_dummies.drop(columns=['unique_id'])

# Print the dataset after removing the 'unique_id' column
print("Dataset after removing the 'unique_id' column:")
print(df_cleaned.head())

Dataset after removing the 'unique_id' column:
     al   su       age    bp       bgr        bu        sc       sod  \
16  0.0  4.0  0.628205  0.50  0.935484  0.058824  0.026087  0.566667   
20  1.0  0.0  0.602564  0.50  0.751613  0.163399  0.078261  0.066667   
27  NaN  NaN  0.525641  0.00  0.067742  0.679739  0.247826  0.733333   
38  NaN  NaN  0.538462  0.25  0.132258  0.888889  0.617391  0.700000   
40  0.0  0.0  0.435897  0.50  0.290323  0.196078  0.139130  0.666667   

         pot      hemo  ...  cad_yes  appet_good  appet_poor  pe_no  pe_yes  \
16  0.500000  0.773913  ...    False        True       False   True   False   
20  0.263158  0.356522  ...    False       False        True   True   False   
27  0.473684  0.200000  ...    False       False        True  False    True   
38  0.578947  0.408696  ...    False        True       False   True   False   
40  0.421053  0.234783  ...    False       False        True   True   False   

    ane_no  ane_yes  Target_ckd  Target_ckd\t

## Part 3.9: Export the Cleaned Data

Now that you've completed these cleaning steps you should have a pandas dataframe which is much cleaner and ready for modeling. Our final step is to save our work. Export the DataFrame to a two new formats: csv and json.

In [12]:
# Export the dataframe to a new csv file
df_cleaned.to_csv('cleaned_dataset.csv', index=False)

# Export the dataframe to a new json file
df_cleaned.to_json('cleaned_dataset.json', orient='records')

# Part 4: Data conversions with Large Language Models

One powerful use case of ChatGPT (and other generative language models) is cleaning and transforming data. In some cases, these models can directly manipulate loosely structured data that you provide to them into a standard format. In the other cases, you can often prompt the model to create a conversion or extraction script for you in python or Pandas and then run it on your own. 

In this part of the assignment you will prompt 383GPT to explore these capabilities.

## Part 4.1 GPT Data Manipulation

Take the cleaned dataset that you created in part three and output the top 15 rows of that dataset. Then copy the terminal output, open 383gpt and ask it to convert that output to a markdown table. Paste that markdown table in the cell bellow

Combined Dataset
| unique_id | al | su | rbc | pc | pcc | ba | htn | dm | cad | ... | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc |
|-----------|----|----|---------|---------|------------|------------|-----|----|-----|-------|------|-------|------|------|-------|-------|------|------|--------|------|
| 102677 | 3.0| 0.0| normal | abnormal | notpresent | notpresent | yes | no | no | ... | 70.0 | 76.0 | 186.0| 15.0 | 135.0 | 7.6 | 7.1 | 22.0 | 3800.0 | 2.1 |
| 103570 | 0.0| 0.0| normal | normal | notpresent | notpresent | no | no | no | ... | 80.0 | 125.0 | NaN | NaN | 136.0 | 3.5 | 15.4 | 43.0 | 5600.0 | 4.5 |
| 104675 | 2.0| 1.0| normal | normal | present | notpresent | no | yes| no | ... | 80.0 | 215.0 | 133.0| 2.5 | NaN | NaN | 13.2 | 41.0 | NaN | NaN |
| 105923 | 1.0| 0.0| NaN | normal | notpresent | notpresent | yes | yes| no | ... | 90.0 | 139.0 | 89.0 | 3.0 | 140.0 | 4.1 | 12.0 | 37.0 | 7900.0 | 3.9 |
| 107297 | 0.0| 0.0| normal | normal | notpresent | notpresent | no | no | no | ... | 70.0 | NaN | NaN | NaN | NaN | NaN | 16.4 | 43.0 | 10800.0| 5.7 |
Dataset After Removing Rows with Missing Values
| unique_id | al | su | rbc | pc | pcc | ba | htn | dm | cad | ... | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc |
|-----------|----|----|---------|---------|------------|------------|-----|----|-----|-------|------|-------|------|------|-------|-------|------|------|--------|------|
| 102677 | 3.0| 0.0| normal | abnormal | notpresent | notpresent | yes | no | no | ... | 70.0 | 76.0 | 186.0| 15.0 | 135.0 | 7.6 | 7.1 | 22.0 | 3800.0 | 2.1 |
| 109053 | 0.0| 0.0| normal | normal | notpresent | notpresent | no | no | no | ... | 60.0 | 106.0 | 27.0 | 0.7 | 150.0 | 3.3 | 14.4 | 42.0 | 8100.0 | 4.7 |
| 114717 | 0.0| 0.0| normal | normal | notpresent | notpresent | no | no | no | ... | 80.0 | 122.0 | 25.0 | 0.8 | 138.0 | 5.0 | 17.1 | 41.0 | 9100.0 | 5.2 |
| 118191 | 0.0| 0.0| normal | normal | notpresent | notpresent | no | no | no | ... | 70.0 | 93.0 | 32.0 | 0.9 | 143.0 | 4.7 | 16.6 | 43.0 | 7100.0 | 5.3 |
| 118191 | 0.0| 0.0| normal | normal | notpresent | notpresent | no | no | no | ... | 70.0 | 93.0 | 32.0 | 0.9 | 143.0 | 4.7 | 16.6 | 43.0 | 7100.0 | 5.3 |
Dataset Sorted by 'Target' Column
| unique_id | al | su | rbc | pc | pcc | ba | htn | dm | cad | ... | bp | bgr | bu | sc | sod | pot | hemo | pcv | wbcc | rbcc |
|-----------|----|----|---------|---------|------------|------------|-----|----|-----|-------|------|-------|------|------|-------|-------|------|------|--------|------|
| 102677 | 3.0| 0.0| normal | abnormal | notpresent | notpresent | yes | no | no | ... | 70.0 | 76.0 | 186.0| 15.0 | 135.0 | 7.6 | 7.1 | 22.0 | 3800.0 | 2.1 |
| 619200 | 0.0| 0.0| NaN | abnormal | notpresent | notpresent | no | yes| no | ... | 80.0 | 171.0 | 30.0 | 1.0 | NaN | NaN | 13.7 | 43.0 | 4900.0 | 5.2 |
| 617089 | 1.0| 3.0| normal | abnormal | notpresent | notpresent | no | no | no | ... | 80.0 | 182.0 | 391.0| 32.0 | 163.0 | 39.0 | NaN | NaN | NaN | NaN |
| 615697 | 4.0| 0.0| abnormal | abnormal | notpresent | present | no | no | no | ... | 60.0 | 94.0 | 67.0 | 1.0 | 135.0 | 4.9 | 9.9 | 30.0 | 16700.0| 4.8 |
| 608902 | 4.0| 1.0| normal | abnormal | present | notpresent | yes | yes| no | ... | 90.0 | 176.0 | 309.0| 13.3 | 124.0 | 6.5 | 3.1 | 9.0 | 5400.0 | 2.1 |
Dataset After Removing Outliers from Numerical Columns
| unique_id | al | su | age | bp | bgr | bu | sc | sod | pot | ... | cad_yes |

### Paste the markdown table here

** Caution: ** while language models can perform data conversions they also can * hallucinate * during this process, particularly for bigger datasets. Reflect on this below, how could you mitigate data conversion hallucinations from LLM conversions?

## Part 4.2 GPT Pandas Prompting

In this section, you will prompt 383GPT to write pandas code manipulations for you.

After working with this data for awhile, we realized we're starting to forget the meanings of the abbreviated column names. Let's ask 383GPT to fix this for us. First, navigate to the [UCI dataset overview](https://archive.ics.uci.edu/dataset/336/chronic+kidney+disease) and copy the abbrevation to name mapping. Then, go to 383GPT and instruct the LLM to provide you with a pandas script to apply this renaming to all the columns of your dataset. Paste that code below and make any adjustments necessary to run it in your notebook.

In [14]:
# Code to rename all the columns in the dataset
import pandas as pd

# Assuming your DataFrame is named 'df'
# Replace this with reading your actual dataset
# df = pd.read_csv('your_dataset.csv')

# Create a dictionary for renaming
renaming_dict = {
    'age': 'age',
    'bp': 'blood pressure',
    'sg': 'specific gravity',
    'al': 'albumin',
    'su': 'sugar',
    'rbc': 'red blood cells',
    'pc': 'pus cell',
    'pcc': 'pus cell clumps',
    'ba': 'bacteria',
    'bgr': 'blood glucose random',
    'bu': 'blood urea',
    'sc': 'serum creatinine',
    'sod': 'sodium',
    'pot': 'potassium',
    'hemo': 'hemoglobin',
    'pcv': 'packed cell volume',
    'wc': 'white blood cell count',
    'rc': 'red blood cell count',
    'htn': 'hypertension',
    'dm': 'diabetes mellitus',
    'cad': 'coronary artery disease',
    'appet': 'appetite',
    'pe': 'pedal edema',
    'ane': 'anemia',
    'class': 'class'
}

# Renaming the columns
df_cleaned.rename(columns=renaming_dict, inplace=True)

# Displaying the DataFrame to check the new column names
print(df_cleaned.head())


    albumin  sugar       age  blood pressure  blood glucose random  \
16      0.0    4.0  0.628205            0.50              0.935484   
20      1.0    0.0  0.602564            0.50              0.751613   
27      NaN    NaN  0.525641            0.00              0.067742   
38      NaN    NaN  0.538462            0.25              0.132258   
40      0.0    0.0  0.435897            0.50              0.290323   

    blood urea  serum creatinine    sodium  potassium  hemoglobin  ...  \
16    0.058824          0.026087  0.566667   0.500000    0.773913  ...   
20    0.163399          0.078261  0.066667   0.263158    0.356522  ...   
27    0.679739          0.247826  0.733333   0.473684    0.200000  ...   
38    0.888889          0.617391  0.700000   0.578947    0.408696  ...   
40    0.196078          0.139130  0.666667   0.421053    0.234783  ...   

    cad_yes  appet_good  appet_poor  pe_no  pe_yes  ane_no  ane_yes  \
16    False        True       False   True   False    True    F

## Part 4.3 Augmenting our skills with prompting

In addition, we can also use 383GPT to convert our data manipulation operations between different data manipulation languages and libraries. For example let's prompt 383GPT to convert the following SQL query to a pandas query.

**SQL Query**
```sql
SELECT Target, COUNT(*) AS count
FROM your_table_name
GROUP BY Target;
```

Prompt 383GPT to convert this to a pandas query. Run this query below, then describe what it does. (If you're not familiar with SQL that is okay you need to only comment on the final resulting output.)

In [25]:
# Verify the 'Target' column is created and display the first few rows

result = cleaned_df.groupby('Target').size().reset_index(name='count')

# Display the result
print(result)

   Target  count
0     ckd     48
1  notckd    150


*Describe what the above code does here*

1) cleaned_df.groupby('Target'): This part of the code groups the rows in the cleaned_df DataFrame by the values in the 'Target' column. The groupby() function essentially creates groups based on the unique values in the 'Target' column. Each group contains rows that share the same value for this column (e.g., 'ckd' and 'notckd').

2) .size(): After grouping the rows by 'Target', the .size() function counts the number of rows (or occurrences) in each group. This produces a Series where the index is the value of 'Target', and the corresponding value is the count of rows for that target value.

3) .reset_index(name='count'): The reset_index() function converts the resulting grouped data back into a DataFrame. It creates a new DataFrame where each unique value from the 'Target' column becomes a row, and the count of occurrences for each value is stored in a new column named 'count'.

4) print(result): This prints the final DataFrame, where each row contains a unique value of 'Target' and its corresponding count, showing how many times each target value appears in the dataset.

